velox/docs/develop/expression-evaluation.rst

```diff
@@ -185,6 +185,45 @@ reordering during execution of the AND and OR expressions.
   :width: 600
   :align: center
 
+Flatten concat-like functions
+`````````````````````````````
+
+Functions that behave like associative operators can declare support for
+flattening. In that case, adjacent nodes of the same function are
+consolidated into one.
+
+A good example is concat(varchar,..) Presto function. Evaluating concat(a, b, c,
+d) is more efficient than evaluating concat(a, concat(b, concat(c, d))).
+Concatenating 4 columns at once allows to calculate the total amount of memory
+needed for the final result, allocate it in one chunk, then copy individual
+values to the right offsets. This saves on memory allocations and reduces data
+copy as compared to concatenating two columns at a time.
+
+Concat Presto function declares support for flattening allowing the expression
+compiler to convert concat(a, concat(b, concat(c, d))) expression to
+concat(a, b, c, d).
+
+Other functions that can leverage this optimization include concat(array,..) and
+map_concat(map,..).
+
+A function declaring support for flattening must have a signature with variadic
+arguments of the same type and return type must be the same as input type.
+
+        f(x,..) -> x
+
+Flattening converts sub-expressions like f(x1, f(x2, f(x3, x4))) into
+f(x1, x2, x3, x4).
+
+Flattening happens before constant folding, hence, f(a, f(constant-x, constant-y))
```

Comment:
Is there any reason for this ? Wouldn't it be better to do constant folding before ?

Reply:
That's a good question. The code is currently structured so that constant folding happens at the very end. I think it is possible to move constant folding before flattening. This would also benefit stateful vector functions which pre-process constant inputs. Currently, this is possible only if inputs are literals, but not if they are constant expressions. CC: @kagamiori